h5fortran not building with pruning enabled

Copied from <https://github.com/fortran-lang/fpm/pull/676#issuecomment-1146772486>:

@LKedward , I found out, that `fpm-0.6.0` doesn't compile correctly
[geospace-code/h5fortran](https://github.com/geospace-code/h5fortran) (it is compilable in `fpm-0.5.0`).
```sh
char_repeat_read                       failed.
[   5%]Compiling...
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/example_char_repeat_read.f90.o: in function `MAIN__':
h5fortran/example/char_repeat_read.f90:22: undefined reference to `__h5fortran_MOD_h5write_scalar'
/usr/bin/ld: h5fortran/example/char_repeat_read.f90:31: undefined reference to `__h5fortran_MOD_h5read_scalar'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x60): undefined reference to `__h5fortran_MOD_h5read_1d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x70): undefined reference to `__h5fortran_MOD_h5read_3d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x78): undefined reference to `__h5fortran_MOD_h5read_2d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x88): undefined reference to `__h5fortran_MOD_h5read_4d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x98): undefined reference to `__h5fortran_MOD_h5read_6d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xa0): undefined reference to `__h5fortran_MOD_h5read_scalar'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xa8): undefined reference to `__h5fortran_MOD_h5write_1d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xb0): undefined reference to `__h5fortran_MOD_h5write_3d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xb8): undefined reference to `__h5fortran_MOD_h5write_4d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xc0): undefined reference to `__h5fortran_MOD_h5write_2d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xc8): undefined reference to `__h5fortran_MOD_h5read_7d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xd0): undefined reference to `__h5fortran_MOD_h5read_5d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xe0): undefined reference to `__h5fortran_MOD_h5write_6d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0xf0): undefined reference to `__h5fortran_MOD_h5write_scalar'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x168): undefined reference to `__h5fortran_MOD_h5write_7d'
/usr/bin/ld: build/gfortran_ECCF645F41985A44/h5fortran/libh5fortran.a(src_interface.f90.o):(.data.rel.ro+0x170): undefined reference to `__h5fortran_MOD_h5write_5d'
collect2: error: ld returned 1 exit status
<ERROR> Compilation failed for object " char_repeat_read "
<ERROR>stopping due to failed compilation
STOP 1
```

Taking `h5write_scalar` as an example, it seems that **[the `submodule` file](https://github.com/geospace-code/h5fortran/blob/main/src/write_scalar.f90)** is removed from the dependency tree.

ENV: Ubuntu, fpm-0.6.0 .

--- a/ci/run_tests.sh
+++ b/ci/run_tests.sh
@@ -76,8 +76,6 @@ test ! -e ./build/gfortran_*/submodule_tree_shake/src_parent_unused.f90.o
 test ! -e ./build/gfortran_*/submodule_tree_shake/src_parent_unused.f90.o.log
 test ! -e ./build/gfortran_*/submodule_tree_shake/src_child_unused.f90.o
 test ! -e ./build/gfortran_*/submodule_tree_shake/src_child_unused.f90.o.log
-test ! -e ./build/gfortran_*/submodule_tree_shake/src_grandchild.f90.o
-test ! -e ./build/gfortran_*/submodule_tree_shake/src_grandchild.f90.o.log
 popd
 
 pushd version_file

--- a/example_packages/submodule_tree_shake/src/child1.f90
+++ b/example_packages/submodule_tree_shake/src/child1.f90
@@ -10,7 +10,7 @@ end interface
 contains
 
 module procedure my_sub1
-    a = 1
+    a = my_fun()
 end procedure my_sub1
 
 end submodule child1

--- a/src/fpm_source_parsing.f90
+++ b/src/fpm_source_parsing.f90
@@ -614,6 +614,7 @@ function parse_sequence(string,t1,t2,t3,t4) result(found)
         select case(token_n)
         case(1)
             incr = len(t1)
+            if (pos+incr-1>n) return
             match = string(pos:pos+incr-1) == t1
         case(2)
             if (.not.present(t2)) exit

--- a/src/fpm_targets.f90
+++ b/src/fpm_targets.f90
@@ -615,24 +615,26 @@ subroutine prune_build_targets(targets, root_package)
         end if
 
         if (allocated(target%source)) then
-            do j=1,size(target%source%modules_used)
-                        
-                if (.not.(target%source%modules_used(j)%s .in. modules_used)) then
 
-                    modules_used = [modules_used, target%source%modules_used(j)]
+            ! Add modules from this target and from any of it's children submodules
+            do j=1,size(target%source%modules_provided)
 
-                    ! Recurse into child submodules
-                    do k=1,size(targets)
-                        if (allocated(targets(k)%ptr%source)) then
-                            if (targets(k)%ptr%source%unit_type == FPM_UNIT_SUBMODULE) then
-                                if (target%source%modules_used(j)%s .in. targets(k)%ptr%source%parent_modules) then
-                                    call collect_used_modules(targets(k)%ptr)
-                                end if
-                            end if
-                        end if
-                    end do
+                if (.not.(target%source%modules_provided(j)%s .in. modules_used)) then
+
+                    modules_used = [modules_used, target%source%modules_provided(j)]
 
                 end if
+
+                ! Recurse into child submodules
+                do k=1,size(targets)
+                    if (allocated(targets(k)%ptr%source)) then
+                        if (targets(k)%ptr%source%unit_type == FPM_UNIT_SUBMODULE) then
+                            if (target%source%modules_provided(j)%s .in. targets(k)%ptr%source%parent_modules) then
+                                call collect_used_modules(targets(k)%ptr)
+                            end if
+                        end if
+                    end if
+                end do
 
             end do
         end if

